### PR TITLE
add support for webpack 5

### PR DESCRIPTION
**Pull request by martinbroos**
Following this guide to allow support for webpack 5 and optimize webpack config files.
https://webpack.js.org/migrate/5/

Until now breaking changes might be solved by checking the wepback version. In this case we could support both versions for now which might be worth it as some project specific plugins might not have webpack 5 support yet.

Already some changes needed to be done at application level. Not sure if we should provide a upgrade guide in this repo or just reference to the guide above when releasing.

todo

- [ ] use css-minimizer plugin instead of cssnano postcss plugin
- [ ] fix remaining typescript errors
- [ ] document some freshheads common used implementations which need upgrade 
- [ ] fix error while building css: `Module build failed (from ../../../webpack-config-builder/node_modules/mini-css-extract-plugin/dist/loader.js):
TypeError: The 'compilation' argument must be an instance of Compilation`

### CODE REVIEW

**Comment by martinbroos** [comment]
@gnoesiboe ik moet nog wel kijken naar de typescript errors. Webpack 5 heeft nu zijn eigen types maar die naamgeving is anders dan wat in @types/webpack gebruikt werd.

File: lib/adapter/ModeAdapter.ts [hunk 1]
```diff
@@ -3,8 +3,7 @@ import { Configuration } from 'webpack';
 import { BuilderConfig } from '../Builder';
 import { warn } from '../utility/messageHelper';
 
-// @todo webpack typescript configuration does not supply a type for this. Create a PR for to be able to re-use it.
```
Comment (gnoesiboe) [inline lib/adapter/ModeAdapter.ts:6]: Fijn dat deze ook weg is nu!

File: lib/adapter/freshheads/JavascriptMinimizationAdapter.ts [hunk 2]
```diff
@@ -2,36 +2,26 @@ import { Adapter, NextCallback } from '../Adapter';
...
+        // Webpack 5.x comes with terser-webpack-plugin so we should be able to require it
         const TerserPlugin = require('terser-webpack-plugin');
 
         new OptimizationAdapter({
             minimize: true,
             minimizer: [
-                new TerserPlugin({
-                    sourceMap: true,
-                }),
+                new TerserPlugin(),
```
Comment (gnoesiboe) [inline lib/adapter/freshheads/JavascriptMinimizationAdapter.ts:24]: Moet `sourcemap` niet meer op `true` gezet te worden?

**Review by gnoesiboe — APPROVED** [review]
Goed bezig! Ruimt de missing type todos ook lekker op!

File: lib/adapter/freshheads/JavascriptMinimizationAdapter.ts (hunk 2 above)
Comment (martinbroos) [reply lib/adapter/freshheads/JavascriptMinimizationAdapter.ts:24]: > 5.0.0 breaking change:
> removed the sourceMap option (respect the devtool option by default)

Comment (gnoesiboe) [reply lib/adapter/freshheads/JavascriptMinimizationAdapter.ts:24]: Goede zet ook!

**Comment by martinbroos** [comment]
Ik merge dit alvast, overige todo's probeer ik vrijdag op te lossen bij de techdag.

**Pull request merged**